[1단계 - 출석 구현] 피글렛(김은수) 미션 제출합니다.

rladmstn opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->

## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  - https://prolog.techcourse.co.kr/studylogs/3958

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
<!-- 이유 작성 -->
모든 원시값 및 문자열을 포장하고자 노력했지만 제대로 된 지 잘 모르겠습니다. 여러 값을 가져와 메서드를 호출하다보니 한 줄에 점을 하나만 찍는다는 사항은 많이 지키지 못했습니다. getter는 최대한 지양하고자 했지만, 해당 값을 가져와 출력하기 위해서 추가하게 되었습니다.

## 어떤 부분에 집중하여 리뷰해야 할까요?

<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 피드백받고 싶은 내용을 공유해주세요.  
예를 들어, 가장 고민했던 점이나 여전히 어려운 부분, 그리고 이에 대한 생각을 적을 수 있습니다. -->
TDD를 집중적으로 해보려고 노력했으나, 여전히 TDD 적용이 어려운 것 같습니다.

예를 들자면, 
1. 출석 확인 기능에 대한 controller 테스트를 먼저 작성한 후에, 작은 부품 메서드들에 대해 테스트를 작성하며 쪼개나가야 할지
2. 도메인들을 미리 설계하고 해당 도메인이 가진 작은 기능들을 먼저 테스트를 생성하며 만들어간 후에 전체 로직을 위해 조립 해야 할지 

이런 부분에서 TDD 적용에 대한 갈피를 잡기가 어려웠던 것 같습니다.
리뷰어 분께서는 어떻게 생각하시는지, TDD 사이클 및 범위를 어떻게 잡으면 좋을지 피드백 해주시면 좋을 것 같습니다!

rladmstn committed:
docs: README 작성

rladmstn committed:
docs: README 작성

rladmstn committed:
docs: 출석 csv 추가

rladmstn committed:
feat: Main 클래스 추가

rladmstn committed:
feat(FileReader): 파일 읽어오는 유틸 클래스 구현

rladmstn committed:
feat(AttendanceChecker): 출석 상태 지정 로직 구현

rladmstn committed:
feat(Attendance): 출석부 구현

출석부를 만들어, 해당 출석부에 없는 이름이 들어온다면 예외를 발생합니다.

rladmstn committed:
feat: 인프라 구축

rladmstn committed:
refactor(AttendanceChecker): 유틸 클래스로 변경

rladmstn committed:
refactor: 도메인 이름 변경

Attendance -> AttendanceBook
AttendanceRecord -> AttendanceRepository
Record -> Attendance

rladmstn committed:
feat(AttendanceBook): 출석부 생성 구현

rladmstn committed:
refactor(AttendanceReader): 출석 초기화 로직 수정

기존에는 파일 리더를 통해 출석 기록만 가져오던 로직을 DTO를 활용하여 출석부와 출결 기록을 함께 가져옵니다.
그에따른 파일들도 수정합니다.

rladmstn committed:
refactor: 컨벤션 진행

rladmstn committed:
feat(Attendance): 출결 기록 비교 로직 구현

비교 조건
1. 닉네임
2. 출석 날짜
같은 것이 있다면 true를 반환하고, 없다면 false를 반환합니다.

rladmstn committed:
feat(AttendanceRepository): 출결 기록 추가 로직 구현

rladmstn committed:
feat: 출석 확인 기능 구현

rladmstn committed:
docs: README 최신화

rladmstn committed:
feat(AttendanceChecker): 주말 출석 검증 로직 구현

rladmstn committed:
feat(AttendanceChecker): 출석 가능 시간 검증 구현

rladmstn committed:
feat(AttendanceRepository): 출석 수정 기능 구현

rladmstn committed:
feat(AttendanceController): 출석 수정 기능 구현

컨트롤러에서 제어하도록 합니다.

rladmstn committed:
refactor(Attendance): 날짜 검증 및 출결 상태 반환 로직 책임 이동

AttendanceChecker의 책임을 Attendance로 이동합니다.

rladmstn committed:
refactor(Time): 시간 객체 일급 컬렉션 적용

3,4번 기능을 위해 LocalDate를 가진 Time 객체를 생성합니다.
그에 따른 수정사항들도 커밋합니다.

rladmstn committed:
fix(Attendance): 기능 요구사항 반영

기존에는 0~5분 지각, 6분부터 결석으로 처리하였습니다.
이제는 ~5분까지 출석, ~30분까지 지각으로 처리합니다.

rladmstn committed:
feat: 3번 기능 정렬 추가 및 존재하지 않는 출석 초기화

1. 3번 기능을 위해, 존재하지 않는 출석을 List에 추가합니다.
2. 3번 기능에 대해서 출력하기 전 정렬 후 리턴합니다.

rladmstn committed:
test(AttendanceRepository): 테스트 수정

rladmstn committed:
fix(OutputView): 출석 수정 결과 출력 오류 수정

rladmstn committed:
feat(AttendanceRepository): 크루 네임에 따른 크루 학적 상태 구하는 로직 구현

출석, 결석, 지각의 수를 구하고 그에 따른 학적 상태를 반환하는 로직을 구현합니다.

rladmstn committed:
feat: 3번 기능 컨트롤러에 추가

rladmstn committed:
feat: 4번 기능 구현

제적 위험자들을 출력하는 기능을 구현합니다.

rladmstn committed:
feat: 재입력 구현 및 버그 수정

1. 시간 수정 안되던 버그 수정합니다.
2. 기능을 입력받을 때 int로 받던 것을 String으로 받습니다.
3. 개행을 추가합니다.
4. 시간 객체를 생성할 때 올바른 시각과 분을 입력하도록 합니다.

rladmstn committed:
refactor(AttendanceController): 입력 메서드 분리

rladmstn committed:
refactor: Time 레코드로 변경 및 Parser 유틸 클래스 생성

1. Time 클래스를 레코드로 변경합니다.
2. Parser 유틸 클래스를 생성합니다.

rladmstn committed:
refactor(AttendanceRepository): 메서드 리팩토링

rladmstn committed:
refactor: 출결 상태 및 학적 상태 ENUM 추가

출결 상태 및 학적 상태를 ENUM으로 바꾸면서
해당 값들을 계산하는 로직을 해당 ENUM들로 리팩터링합니다.

rladmstn committed:
refactor: 파라미터 final 추가

rladmstn committed:
docs: README 진행 사항 업데이트 및 공통 요구사항 수정

rladmstn committed:
fix(AttendanceStatus): 오타 수정

rladmstn committed:
refactor(Time): 등교 가능 시간 검증 메서드 리팩터링

등교 가능 날짜 및 시간 검증을 한 번에 하던 validatePossibleTime 메서드를
등교 가능 날짜 검증 메서드와 등교 가능 시간 검증 메서드로 분리하였습니다.

jamie9504 commented on README.md:
```diff
@@ -1,3 +1,44 @@
 # java-attendance
```
1. > 여러 값을 가져와 메서드를 호출하다보니 한 줄에 점을 하나만 찍는다는 사항은 많이 지키지 못했습니다. 

    - 한 줄에 점을 하나만 찍는다는 것을 왜 지켜야 하는지 알아보고, 간단히 정리해보세요.

2. >  getter는 최대한 지양하고자 했지만, 해당 값을 가져와 출력하기 위해서 추가하게 되었습니다.

    - getter를 지양해야 하는 이유에 대해 알아보고, 간단히 정리해보세요.

3. > 출석 확인 기능에 대한 controller 테스트를 먼저 작성한 후에, 작은 부품 메서드들에 대해 테스트를 작성하며 쪼개나가야 할지, 도메인들을 미리 설계하고 해당 도메인이 가진 작은 기능들을 먼저 테스트를 생성하며 만들어간 후에 전체 로직을 위해 조립 해야 할지, 이런 부분에서 TDD 적용에 대한 갈피를 잡기가 어려웠던 것 같습니다. 리뷰어 분께서는 어떻게 생각하시는지, TDD 사이클 및 범위를 어떻게 잡으면 좋을지 피드백 해주시면 좋을 것 같습니다!

    - Outside-In 접근 방식(하향식: ex. controller부터)과 Inside-Out 접근 방식(상향식: ex. domain부터) 중 무엇이 더 좋은지는 흔히 하는 고민이죠!
    - 개발할 때 하나만 사용할지, 혼용해서 사용할지, 어떤 것이 적재적소에 사용될지는 결국 장/단점을 비교해보아야하는데요.
    - [숙제] 각 접근 방식의 장/단점을 고민해보고, 페어가 아닌 크루 3명과 자신들이 생각하는 장/단점과 각자가 선호하는 방식과 적용하면 좋을 부분에 대해 논의해보세요. 논의 후 논의 내용과, 피글렛의 논의 전/후의 생각을 적어주세요!

rladmstn replied:
**1. 한 줄에 점을 하나만 찍는 다는 것은 두 가지의 의미가 있을 것이라고 생각했습니다.**

첫 번째는 메서드 체이닝을 피하기 위함인데, 하나의 객체에 대해서 여러 개의 메서드들로 이어질 경우 각 메서드가 어떤 역할을 하는지 파악하기도 어렵고 예외 발생시의 원인도 추적하기 어렵다고 생각합니다. 

두 번째로는 객체가 다른 객체의 내부 구조나 상태를 알지 못하도록, 결합도를 낮추기 위해 존재하는 디미터 법칙을 지키기 위함이라 생각했습니다. 

처음에는 stream() 사용에 의한 점도 위 법칙에 위배된다고 생각했었으나, 위 원칙에 대해 코드를 보며 생각을 정리해보니 stream()은 위 법칙과 거리가 먼 사항이라는 생각이 들었습니다. 다시 생각해볼 수 있던 기회였습니다!

---

**2. getter를 지양해야 하는 이유는 객체의 캡슐화와 변경 가능성 때문이라고 생각합니다.**

캡슐화 자체가 객체의 내부 상태를 외부에서 접근하지 못하게 하고, 메서드로 데이터를 조작하도록 만드는 개념입니다. 하지만 getter를 사용하게 되면 캡슐화도 약해져 외부에서 해당 데이터를 조작할 수 있게 되어 좋지 못한 설계가 된다고 생각합니다.

이러한 부분을 어떻게 보완할 수 있을까 고민했는데, 불변 객체 및 새로운 객체를 반환하면 getter를 사용하고도 캡슐화를 더 지킬 수 있는 방법이 되리라 생각해 이번 리팩토링에서 적용해보았습니다.

---
**3. 데일리 미팅을 함께하는 크루들과 다같이 의논해보았습니다!**

크루들 중 대다수가 설계는 Top-down으로 한 후에 설계한 도메인들을 바탕으로 Bottom-up 형태의 TDD를 적용했다고 합니다. Bottom-up으로 TDD를 진행하면 도메인 설계 기반으로 TDD를 할 수 있기 때문에 더 꼼꼼하고 확실한 TDD 사이클을 돌릴 수 있었다는 점이 장점이라는 의견이 있었습니다. 하지만, 그만큼 도메인 설계를 하는데 시간이 걸려서 초기 설계에 시간을 많이 쏟게 된다는 단점이 있다고 합니다. 

Top-down 방식으로 TDD를 적용하면 개발을 시작하기 위한 초기 시간이 짧다는 점에서는 장점이 될 수 있지만, 그 후에 쪼개지는 도메인들로 인해 테스트와 프로덕트 코드에 대해 리팩토링 횟수가 잦아진다는 단점이 있었습니다.

저는 이번 미션에서는 거의 Bottom-up 형식의 TDD를 진행했던 것 같습니다만, 중간중간 domain의 역할을 더 쪼갤 수 있는 지점이 발생하면 다시 Top-down의 형태가 됐던 것 같습니다. 저 또한 설계의 틀이 꼼꼼하게 진행된다면 Bottom-up의 TDD가 나을 것이라고 생각이 들었었는데, 크루들과 논의해보니 하나의 방법으로는 효율적인 개발이 되지 못할 것이라고 생각이 들어요. 둘의 방식은 적재적소 맞게 어느정도 혼용되어야 할 방식들이라고 생각이 듭니다!

jamie9504 replied:
1. 의견 정리 좋네요!
stream을 한 줄에 모두 선언했을 때와 나누었을 때 디버깅 포인트를 잡는 것은 어떠한 차이가 있고, 어떤 것이 편할지도 고민해보세요.

2. 좋아요!
[숙제] 불변이 아닌 객체도, getter를 써야할 때가 있을텐데 이럴 땐 어떻게 해야 좋을지, 무조건 불변을 강제화해야 좋을지 데일리 미팅 크루들과 이야기를 나눠보세요!

3. 맞습니다. Top-Down과 Bottom-Up 중 하나만 사용하는 것이 항상 정답은 아니에요. 하이브리드로 필요할 때 적재적소에 맞춰서 사용하는 것이 중요합니다. 어떨 때 무엇을 쓰는 것이 좋은가는 TDD가 익숙해지면서 생각이 계속 바뀔 순 있어요. 정답은 없습니다. 피글렛만의 방식을 찾아가보세요!

rladmstn replied:
> 1. 의견 정리 좋네요!
>    stream을 한 줄에 모두 선언했을 때와 나누었을 때 디버깅 포인트를 잡는 것은 어떠한 차이가 있고, 어떤 것이 편할지도 고민해보세요.

stream() 또한 제가 생각한 1번의 의견에서 해당될 수도 있겠네요! stream()도 한 줄에 선언했을 때와 여러 줄에 걸쳐서 선언했을 때, 어떤 부분에서 예외가 발생했는지 판단할 수 있으려면 한 번에 모두 선언하는 것보다 여러 줄에 걸쳐서 선언하는 것이 break point를 걸 때도 유용할 것 같습니다! 

---

> 2. 좋아요!
>    [숙제] 불변이 아닌 객체도, getter를 써야할 때가 있을텐데 이럴 땐 어떻게 해야 좋을지, 무조건 불변을 강제화해야 좋을지 데일리 미팅 크루들과 이야기를 나눠보세요!

이번에도 크루들과 함께 이야기 해보았는데요! 변경 가능한 객체에 대해서 getter를 사용하게 된다면 방어적 복사를 사용하도록 만든다는 의견이었습니다. 원본 참조는 가능하지만 변경은 불가능하게 하는 `unmodifiableList()`와 같은 것들을 사용한다고 하더라구요. 저 또한 변경이 가능한 객체도 getter를 쓸 땐 방어적 복사를 해야한다고 생각합니다! 

하지만, 무조건 getter를 쓸 때 불변을 강제화하는 것은 또다른 GC의 대상이 될 것이라고 생각해요. getter로 정보만 가져오기 위해 한 번 쓴 후에 재사용 되지 못하는 객체라면 이는 곧 GC에서 처리해야 할 수거 대상이 되리라 생각합니다.

getter에서 불변성을 강제화 하는 것도, 상황에 따라서 정보의 변경이 민감하게 여겨질 경우에 지켜지면 좋지 않을까라는 생각이 있습니다 :) 해주신 질문에 대해서는 제이미의 의견도 들어보고 싶습니다..!

---

> 3. 맞습니다. Top-Down과 Bottom-Up 중 하나만 사용하는 것이 항상 정답은 아니에요. 하이브리드로 필요할 때 적재적소에 맞춰서 사용하는 것이 중요합니다. 어떨 때 무엇을 쓰는 것이 좋은가는 TDD가 익숙해지면서 생각이 계속 바뀔 순 있어요. 정답은 없습니다. 피글렛만의 방식을 찾아가보세요!

감사합니다! 이번 미션 수행하면서 TDD를 체화할 수 있는 가장 좋은 방법이 무엇일지 꾸준히 고민 해야겠어요 :)

jamie9504 replied:
오 많은 고민을 또 하셨군요!

불변에 대한 제 입장은.. 도메인에 따라 협업하는 사람에 따라 매번 다르지만,
팀에서 불변을 명시적으로 사용하고 있는가, 아니면 묵시적으로 불변이 아니더라도 불변처럼 취급하고 있는가.
변경하는 포인트를 팀 룰에서 제어할 수 있는가 없는가에 따라 다릅니다 :)

불변을 선호한다면, 불변을 선호하지 않는 개발자와 어떻게 협업을 해야할지
불변을 선호하지 않는다면, 불변을 선호하는 개발자와 어떻게 협업을 해야할지 고민해보는 것이 시야를 넓히는 데 도움이 될 것 같아요.

jamie9504 commented on src/main/java/attendance/Main.java:
```diff
@@ -0,0 +1,18 @@
+public class Main {
+
+    public static void main(String[] args) {
+
```
메서드별로 첫 줄 개행이 다른데요. 일관성있게 맞춰주세요.
두 줄 이상의 불필요한 개행도 체크해주세요.

rladmstn replied:
개행 문자도 일종의 컨벤션이 될 수 있는데, 이번 미션을 진행하면서 제대로 맞추지 못한 것 같네요..! 확인하고 정리해보겠습니다!

jamie9504 commented on src/main/java/attendance/controller/AttendanceController.java:
```diff
@@ -0,0 +1,155 @@
+    }
+
+    private boolean choiceFunction(final String functionValue) {
+        if (functionValue.equals("1")) {
```
> Java Enum을 적용한다.

가능하지 않을까요?

rladmstn replied:
처음에 enum을 적용해보려 했었는데, enum으로 갖고 있을 때의 장점을 판단하지 못했어서 다시 제거하게 되었습니다. enum으로 function을 가졌을 때의 장점을 다시 판단해보고 적용해보도록 하겠습니다!

jamie9504 replied:
Enum 사용 관련 보충 설명) Controller에서 1이 어떤 의미인지 알아야 할 필요가 있을까요?

rladmstn replied:
controller는 view에서 입력받는 값을 바탕으로 domain과의 사이를 연결해주는 역할이라고 생각합니다. 따라서 enum 내에서 어떤 기능이 어떤 문자값과 매칭 되는지만 관리해주면 controller는 입력과 기능 간 매핑 관계를 알 필요가 없을 것이라는 생각이 듭니다! 추가로 고민해볼 기회 주셔서 감사합니다!!

jamie9504 commented on src/main/java/attendance/controller/AttendanceController.java:
```diff
@@ -0,0 +1,155 @@
+    public void start() {
+
+        initAttendanceSystem();
+        do {
```
> indent(인덴트, 들여쓰기) depth를 2를 넘지 않도록 구현한다. 1까지만 허용한다.
> 예를 들어 while문 안에 if문이 있으면 들여쓰기는 2이다.
> 힌트: indent(인덴트, 들여쓰기) depth를 줄이는 좋은 방법은 함수(또는 메서드)를 분리하면 된다.

인덴트 체크해주세요.

rladmstn replied:
놓쳤던 부분인데 캐치 감사합니다! :)

jamie9504 commented on src/main/java/attendance/domain/AcademicStatus.java:
```diff
@@ -0,0 +1,40 @@
+        return value;
+    }
+
+    public static String getAcademicStatus(int late, int absent) {
```
Enum을 반환하는 방식과 Enum의 value까지 반환하는 방식은 어떠한 차이가 있을까요?

rladmstn replied:
enum을 반환하게 되면 다른 비즈니스 로직에서도 해당 반환 값을 사용할 때, 요구사항 및 필요성에 따라 `getValue()`를 호출해 String으로 쓸 수도 있고 비교를 위해 `==`를 사용할 수 있다는 장점이 있는 것 같습니다.
그에 반해 value를 반환하는 방법은 단순 출력에만 필요할 메서드일 경우에만 편히 사용될 수 있을 뿐더러, 반환 값을 통한 비교의 경우 `equals()`를 사용해야 하므로 하드코딩의 실수 가능성과 가독성 하락의 문제점이 있다고 생각합니다.

처음 위 메서드를 만들 땐, 해당 메서드의 사용처가 한 곳이고 어차피 value를 사용해야 했기에 위처럼 작성했었습니다. 하지만 달아주신 코멘트를 보고 다시 생각해보니, 유지보수성과 확장성을 위해서라면 value보단 enum으로 반환하는 것이 더 좋은 방법이라 생각해 수정했습니다! 

생각지 못한 부분이었는데, 다시 한 번 고민해볼 수 있는 기회가 되었습니다! 감사합니다 :)

jamie9504 commented on src/main/java/attendance/domain/Attendance.java:
```diff
@@ -0,0 +1,47 @@
+    public boolean isSameByNameAndLocalDate(final String name, int year, int month, int day) {
+        return crewName.equals(name) && day == attendanceTime.getDay() && year == attendanceTime.getYear()
+                && month == attendanceTime.getMonth();
+    }
```
단위 테스트가 되었나요? 미흡한 부분이 있다면 채워주세요.
TDD 싸이클대로 진행을 했는데, 단위테스트가 미흡했다면 그 이유는 무엇일지 고민해보세요.

rladmstn replied:
Top-down으로 설계하다가 Bottom-up으로 TDD를 진행하니, 중간중간 도메인을 추가해서 책임을 쪼갤 수 있겠다는 생각을 하며 진행하다보니 테스트 작성이 미흡했던 부분이 있었습니다.

다음 스텝에서는 TDD 사이클을 더 신경써서 지켜보도록 노력해봐야 할 것 같아요..!

rladmstn replied:
Bottom-up으로 진행하다가 중간중간 도메인을 추가해서 책임을 쪼갤 수 있겠다는 생각으로 진행하다보니 다시 Top-down 형태가 되면서 테스트 작성이 미흡했던 부분이 있었습니다.

다음 스텝에서는 빠진 테스트가 없는지 더 신경쓰고 TDD 사이클을 더 지켜보도록 노력해봐야 할 것 같아요..!

jamie9504 commented on src/main/java/attendance/domain/AttendanceBook.java:
```diff
@@ -0,0 +1,39 @@
+        this.names = names;
+    }
+
+    public void checkName(final String name) {
```
check~는 어떤 동작을 할 거라고 예상할 수 있을까요? 이름이 모호한 느낌이 있네요.

rladmstn replied:
크루 목록 중에 해당 이름이 있는지를 검증해주는 메서드였는데, 다시 보니 `check~` 보단 `validate~` 형식의 메서드 이름이 검증의 역할을 더 잘 보여줄 것 같아요! 피드백 감사합니다 :)

jamie9504 commented on src/main/java/attendance/domain/AttendanceBook.java:
```diff
@@ -0,0 +1,39 @@
+        return names.stream()
+                .map(attendanceRepository::getAcademicStatusByName)
+                .filter(dto -> dto.academicStatus().equals(academicStatus))
+                .sorted(Comparator.comparing(CrewNameAndAcademicStatusDTO::crewName))
```
domain에서 dto를 사용한다면, 어떤 이슈가 있을 수 있을까요?
[숙제] 고민해보고, 페어가 아닌 크루 3명과 토론해본 후 고민해본 점, 토론한 내용, 토론 후 피글렛의 생각을 정리해주세요.

rladmstn replied:
팀원들과 함께 의견을 나눠본 결과, 공통적으로 공감하는 점은 **DTO가 View와 근접한 객체이기에 변동에 영향을 많이 받을 것**이라는 점입니다.

또한 DTO는 도메인을 보호하며 필요한 정보를 외부로 전달하는 역할이기 때문에, 도메인에서 사용하게 되면 책임이 불분명해진다는 점이 이슈가 될 수 있을 것 같습니다. 내부 도메인 간 교류에서 DTO를 사용하게 되면 캡슐화가 이뤄지지 않는다는 점에서 객체지향적 특성도 잃게 되리라 생각합니다! 

기존에도 DTO의 생각은 위 의견과 동일했지만, 이번 미션을 진행하면서는 미처 고려하지 못했던 것 같습니다. 의견 공유해보며 다시 한 번 DTO의 의미에 대해 고민해볼 수 있던 시간이었습니다!

jamie9504 commented on src/main/java/attendance/domain/Time.java:
```diff
@@ -0,0 +1,76 @@
+import java.time.format.TextStyle;
+import java.util.Locale;
+
+public record Time(LocalDate date, String hour, String minute, boolean isAbsent) {
```
이 객체는 역할이 있는데, Time이라는 이름이 그 역할을 충분히 나타내주고 있나요?

rladmstn replied:
**출석** 시간과 이에 대한 검증 역할을 갖고 있으므로 `Time`보단 `AttendanceTime` 이 더 역할을 잘 보여줄 것 같아요. `AttendanceTime`으로 수정해보겠습니다!

jamie9504 commented on src/main/resources/attendances.csv:
```diff
@@ -0,0 +1,43 @@
+빙티,2025-02-14 13:00
+쿠키,2025-02-14 13:01
+이든,2025-02-14 13:02
+빙봉,2025-02-14 13:06
```
github은 마지막 개행이 없으면 (-) 마크가 뜹니다.

rladmstn replied:
잘 몰랐던 사실인데 그렇네요..! 알려주셔서 감사합니다!!

jamie9504 reviewed (changes requested):
안녕하세요, 피글렛. 리뷰어 제이미입니다.

만나서 반갑습니다.

코멘트 남겨두었으니 확인해주세요.

rladmstn committed:
fix(attendances.csv): 마지막 줄 개행 문자 추가

rladmstn committed:
refactor(Function): 기능 상수 위한 Function enum 추가

사용자가 입력하는 기능 상수들을 관리하는 Function enum을 추가했습니다.

rladmstn committed:
refactor: 개행 문자 일괄 수정

메서드 시작 시 개행 문자를 일괄 추가하도록 수정했습니다.

rladmstn committed:
refactor(Time): Time을 AttendanceTime으로 이름 변경

출석 시간 관련한 기능을 담은 Time의 역할이 더 잘 드러나도록 AttendanceTime 으로 이름을 수정합니다.

rladmstn committed:
refactor(OutputView): 개행 출력 추가 및 포매팅 수정

출력 기능의 마지막 줄에 개행 출력을 추가했습니다.
메세지 포맷팅 방법을 String.format()에서 printf로 수정했습니다.

rladmstn committed:
refactor(AttendanceController): 재입력 로직 수정 및 네이밍 변경

예외적인 입력을 받았을 때, 해당 부분부터 재입력 받도록 수정합니다.
AttendanceBook의 이름 검증 메서드 네이밍을 check에서 validate로 수정합니다.

rladmstn committed:
refactor(AttendanceStatus): 출석 상태 반환하는 메서드 리턴 타입 수정

기존에는 AttendanceStatus의 String value를 반환했었지만, 해당 메서드의 확장성 및 유지보수성을 위해 enum을 반환하도록 수정합니다.

rladmstn committed:
refactor(CrewAttendanceInformation): DTO를 domain으로 변경

기존에 데이터 전달을 위해 CrewNameAndAcademicStatusDTO를 사용했으나,
해당 데이터를 사용한 데이터 정렬 및 가공이 존재해 해당 클래스를 CrewAttendanceInformation 도메인으로 수정했습니다.

rladmstn committed:
refactor(AcademicStatus): 크루 학적 상태 반환하는 메서드 리턴 타입 수정

지각 및 결석 횟수로 크루의 학적 상태를 반환하는 메서드에서 String value 값을 반환했지만,
유지보수성을 위해 enum을 반환하도록 수정합니다.

rladmstn committed:
test(AcademicStatus): 크루 학적 상태 반환 메서드 테스트 작성

rladmstn committed:
test(AttendanceStatus): 출석 상태 반환 메서드 테스트 작성

rladmstn committed:
refactor(Attendance): AttendanceTime getter 복사본 반환

getter에 따른 객체 캡슐화 약화를 보완하기 위해 새로운 객체 복사본을 반환하도록 수정합니다.

rladmstn committed:
refactor(Attendance): 누락 테스트 작성

해당 날짜에 입력받은 크루의 출석 기록이 존재하는지 확인하는 메서드 테스트를 작성했습니다.

jamie9504 commented on src/main/java/attendance/controller/AttendanceController.java:
```diff
@@ -0,0 +1,226 @@
+import java.util.List;
+import java.util.Locale;
+import java.util.function.Supplier;
+
+
```
> 두 줄 이상의 불필요한 개행도 체크해주세요.

rladmstn replied:
헉 이 부분을 제대로 확인 못했네요 ㅜㅜ 다시 꼼꼼하게 정리해보겠습니다..!

jamie9504 commented on src/main/java/attendance/controller/AttendanceController.java:
```diff
@@ -0,0 +1,226 @@
+    private final InputView inputView;
+    private final OutputView outputView;
+
+    private AttendanceBook attendanceBook;
```
Controller에서 AttendanceBook를 가지고 있는데, 어떤 장/단점이 있을까요?

rladmstn replied:
Controller가 AttendanceBook을 갖고 있으면 controller 내 메서드에서 AttendanceBook을 사용할 때 메서드 파라미터로 매번 넘기지 않아도 된다는 장점이 있습니다. 
하지만 Controller는 도메인 간 연결을 해주는 역할을 해야하는데, AttendanceBook을 갖고 있게 되면 직접 관리해야하는 데이터 관리 책임이 추가된다는 단점이 있습니다. 이는 곧 비즈니스 로직의 일부를 담당하게 되어 관심사 분리가 제대로 이뤄지지 않을 수 있다는 우려가 존재한다고 생각합니다.

현재로서는 AttendanceBook에서 이름 검증 역할을 갖고 있고, 이를 위해선 input을 위해 매번 AttendanceBook을 넘겨줘야 하는데, 이러한 과정에서 번거로움을 느껴 Controller가 갖고 있도록 구성했었습니다. 

제이미는 이런 구조에서 번거롭더라도 메서드 파라미터로 AttendanceBook을 계속 넘겨주는 것을 선호하시는지, 아님 지금처럼 Controller에서 갖고있도록 만드는 것을 선호하시는지 궁금합니다!

jamie9504 replied:
Controller에서 상태를 가지고 있다면, 해당 컨트롤러를 여러 곳에서 사용할 수 있을까요?

`Controller는 도메인 간 연결을 해주는 역할을 해야하는데, AttendanceBook을 갖고 있게 되면 직접 관리해야하는 데이터 관리 책임이 추가된다는 단점`이라면, 이 객체를 Controller라고 볼 수 있을지 고민해보면 좋을 것 같아요.

jamie9504 commented on src/main/java/attendance/controller/AttendanceController.java:
```diff
@@ -0,0 +1,226 @@
+    }
+
+    private void runFunction(final Function function) {
+        if (function == Function.ATTEND) {
```
> 메서드별로 첫 줄 개행이 다른데요. 일관성있게 맞춰주세요.

피글렛은 어떠한 기준을 적용하고 있나요?

rladmstn replied:
메서드 시작 시 개행을 한 줄 추가하고 작성합니다!

다시 확인해보니 통일이 안된 부분이 몇 군데 더 있었네요..! 꼼꼼히 확인해봐야겠습니다..

jamie9504 commented on src/main/java/attendance/domain/AcademicStatus.java:
```diff
@@ -0,0 +1,40 @@
+        return value;
+    }
+
+    public static AcademicStatus getAcademicStatus(int late, int absent) {
+
+        return Stream.of(late / 3 + absent)
+                .map(count -> {
+                    if (count > 5) {
+                        return EXPELLED;
+                    }
+                    if (count >= 3) {
+                        return INTERVIEW;
+                    }
+                    if (count == 2) {
+                        return WARNING;
+                    }
+                    return NOT;
+                })
```
피글렛은 어떠한 기준으로 클래스 내 메서드를 선언하시나요?

---

> indent(인덴트, 들여쓰기) depth를 2를 넘지 않도록 구현한다. 1까지만 허용한다.
> 예를 들어 while문 안에 if문이 있으면 들여쓰기는 2이다.
> 힌트: indent(인덴트, 들여쓰기) depth를 줄이는 좋은 방법은 함수(또는 메서드)를 분리하면 된다.

rladmstn replied:
저는 메서드 하나당 하나의 역할만 하도록 나누려고 집중하는 편입니다! 중복을 줄이고 여러 곳에서 재사용 가능한 로직들을 빼두려고 노력하고 있습니다. 또한, 너무 긴 메서드는 작은 단위로 나누어서 가독성을 높이고자 메서드를 선언하려고 합니다.

---

map 내에서 indent가 발생하는 것도 포함되는 이야기인데 미처 생각을 못한 것 같습니다 ㅜㅠ 다시 확인하겠습니다..!

jamie9504 replied:
(추가) 피글렛은 어떠한 기준으로 클래스 내 메서드를 선언하시나요?

ex) public 메서드, private 메서드, getter-메서드, setter-메서드 etc

[숙제] 다음 PR 본문에 지금 사용한 방식을 한 번 정리해보시고, 크루들은 어떻게 하고 있는지 3명과 의견을 나누고 정리해봅시다.

jamie9504 commented on src/main/java/attendance/domain/Attendance.java:
```diff
@@ -0,0 +1,50 @@
+
+    public String getAttendanceStatus() {
+
+        if (attendanceTime.getDayOfWeek().equals("월요일")) {
+            return AttendanceStatus.getAttendanceStatusWithCondition(attendanceTime, 13, 5, 30).getValue();
+        }
```
매직 넘버에 대해 고민해볼까요?

rladmstn replied:
의미있는 숫자들에 대해서는 매직넘버를 클래스 내 상수 처리 하는 것이 좋다고 생각합니다! 미션 진행 동안 많이 적용하지 못한 것이 아쉬웠습니다 :( 이번 피드백에서 적용해보겠습니다!!

jamie9504 commented on src/main/java/attendance/domain/Attendance.java:
```diff
@@ -0,0 +1,50 @@
+
+    public boolean isAlreadyAttendance(final Attendance currentAttendance) {
+
+        if (!crewName.equals(currentAttendance.crewName)) {
+            return false;
+        }
```
이 부분에 대한 단위테스트 되고 있나요?

rladmstn replied:
해당 부분에 대한 테스트가 빠졌네요..! 더 꼼꼼히 확인하고 작성 해야겠어요 감사합니다 !

jamie9504 commented on src/main/java/attendance/domain/AttendanceBook.java:
```diff
@@ -0,0 +1,39 @@
+        }
+    }
+
+    public void initAbsent(final AttendanceRepository attendanceRepository) {
+
+        for (String name : names) {
+            attendanceRepository.initAbsent(name);
+        }
+    }
```
요 메서드는 테스트되지 않아도 괜찮은가요?
전반적으로 테스트코드 한 번 더 확인해주세요~

---

또한 AttendanceBook이 Repository를 알고 있는데, 패키지 순환 참조에 대해서도 알아볼까요?

rladmstn replied:
전체적으로 여기저기 빠진 테스트 코드들이 몇 개 더 있던 것 같아요..! 다시 확인해서 추가해두겠습니다!

---

근원적인 패키지 순환 참조의 문제점을 제대로 알지 못했었는데, 찾아보니 의존 받는 쪽에서 변경이 일어나면 의존 하는 곳에서 연쇄적으로 변경에 대한 영향이 발생한다는 문제점이 있더라구요. 이런 점을 고려해서 현재 `AttendanceRepository`와 `AttendanceBook` 간의 순환 참조가 발생하는 것을 끊어내고자, `AttendanceRepository`가 `AttendanceBook`를 참조하는 단방향이 되도록 수정해보았습니다.
제적 위험자 조회 메서드 자체가 AttendanceRepository의 역할이 되기도 할 것 같아서 repository로 옮겼는데요, Controller에서 AttendanceBook의 names를 얻어와 repository의 이 메서드를 호출하도록 수정했습니다!

+) getter로 Attendance의 정보(필드)를 가져온다는 단점이 있지만, 이는 순환 참조 보다는 강도가 낮은 side effect라고 생각했습니다! `Set.of()`으로 변경 가능성도 막을 수 있기 때문에 충분히 사용할 만 하다고 판단해 이렇게 리팩토링 하게 되었습니다.

jamie9504 replied:
.

jamie9504 commented on src/main/java/attendance/utils/AttendanceReader.java:
```diff
@@ -0,0 +1,41 @@
+
+            String crewName = split[0];
+            LocalDateTime attendanceTime = LocalDateTime.parse(split[1],
+                    DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm"));
```
실행될 때마다 매번 해당 객체를 생성하고 있군요.

rladmstn replied:
`DateTimeFormatter` 말씀하신거라면 미션 진행하면서 놓친 것 같습니다..ㅎㅎ `AttendanceReader`가 static final로 갖고 있어도 충분히 괜찮은 구조가 될 것 같습니다! 캐치 감사합니다 :)

jamie9504 reviewed (changes requested):
안녕하세요, 피글렛. 리뷰어 제이미입니다.

코멘트 조금 더 남겨두었습니다.

rladmstn committed:
refactor: 메서드 시작 시 개행문자 추가 일괄 적용

rladmstn committed:
refactor: 불필요한 개행 정리

rladmstn committed:
refactor: 매직넘버 및 파일 경로 상수화

기존 하드코딩 했던 매직넘버와 파일 경로를 static final로 상수화 했습니다.

rladmstn committed:
refactor(FileReader): 메서드 분리

indent 1 유지를 위해 파일을 읽어오는 부분의 메서드를 분리했습니다.

rladmstn committed:
fix(AttendanceTime): 캠퍼스 운영 종료 시간 상수 수정

캠퍼스 운영 종료 시간인 23시 상수가 잘 못 선언되어 있어 수정합니다.

rladmstn committed:
test: 누락된 테스트 추가

rladmstn committed:
refactor: 컨벤션 따른 개행 및 코드 정리

rladmstn committed:
refactor(AttendanceRepository): 제적 위험자 조회 메서드 이동

기존 AttendanceBook에서 갖고있던 제적 위험자 조회 메서드를 순환참조 방지 및 책임 이전을 위해 AttendanceRepository로 이동합니다.

rladmstn committed:
refactor(AttendanceRepository): 결석 처리 초기화 메서드 이동

기존 AttendanceBook에서 결석 처리를 초기화하던 메서드를,
순환참조 방지 및 책임 이전을 위해 AttendanceRepository로 이동합니다.

rladmstn committed:
test(AttendanceTime): AttendanceTime 검증 순서 수정 및 테스트 추가

AttendanceTime을 생성할 때 시간 검증 로직의 순서를 수정하고, 누락된 예외 테스트를 추가합니다.

rladmstn committed:
refactor(AttendanceController): 파일 경로 상수명 수정

카멜 케이스로 작성했던 파일 경로 상수명을 전체 대문자와 스네이크 케이스를 적용하도록 수정했습니다.

rladmstn committed:
refactor: 상수명 수정 및 네이밍 변경

카멜 케이스로 작성했던 상수명을 전체 대문자와 스네이크 케이스를 적용하도록 수정했습니다.
repository의 initAbsent와 initStudentAbsent의 네이밍이 모호하다고 판단해, initAllAbsent, initCrewAbsent로 수정했습니다.

jamie9504 commented on src/main/java/attendance/domain/AcademicStatus.java:
```diff
@@ -0,0 +1,40 @@
+    EXPELLED("제적"),
+    NOT("X");
+
+    private final String value;
+
+    private static final int EXPELLED_COUNT = 5;
+    private static final int INTERVIEW_COUNT = 3;
+    private static final int WARNING_COUNT = 2;
+    private static final int LATE_AS_ABSENT = 3;
```
상수와 멤버변수 내의 선언 순서 기준이 있나요?

COUNT는 반복되고 있는데, value로 사용하는 것과 상수로 사용하는 건 어떤 차이가 있나요?

jamie9504 commented on src/test/java/attendance/repository/AttendanceRepositoryTest.java:
```diff
@@ -0,0 +1,243 @@
+
+        return attendances;
+    }
+}
```
> github은 마지막 개행이 없으면 (-) 마크가 뜹니다.

jamie9504 commented on src/test/java/attendance/utils/FileReaderTest.java:
```diff
@@ -0,0 +1,26 @@
+import org.junit.jupiter.api.Test;
+
+class FileReaderTest {
+
+
```
> 두 줄 이상의 불필요한 개행도 체크해주세요.

jamie9504 reviewed (approved):
안녕하세요, 피글렛. 리뷰어 제이미입니다.

코멘트 조금 더 남겨두었는데요.
자잘한 것이라 2단계 진행해주시면서 함께 해주시면 됩니다.

관련한 답변은 2단계 PR 본문에 함께 적어주세요.
(1단계에 작성시 확인이 어렵습니다)

이만 Approve & Merge합니다.
